wen-restart/src/last_voted_fork_slots_aggregate.rs

```diff
@@ -3,69 +3,108 @@ use {
     anyhow::Result,
     log::*,
     solana_gossip::restart_crds_values::RestartLastVotedForkSlots,
-    solana_runtime::epoch_stakes::EpochStakes,
-    solana_sdk::{clock::Slot, hash::Hash, pubkey::Pubkey},
+    solana_runtime::bank::Bank,
+    solana_sdk::{
+        clock::{Epoch, Slot},
+        hash::Hash,
+        pubkey::Pubkey,
+    },
     std::{
-        collections::{HashMap, HashSet},
+        collections::{BTreeSet, HashMap, HashSet},
         str::FromStr,
+        sync::Arc,
     },
 };
 
+// If at least 1/3 of the stake has voted for a slot in next Epoch, we think
+// the cluster's clock is in sync and everyone will enter the new Epoch soon.
+// So we require that we have >80% stake in the new Epoch to exit.
+const EPOCH_CONSIDERED_FOR_EXIT_THRESHOLD: f64 = 1f64 / 3f64;
+
+#[derive(Debug, Clone, PartialEq)]
+pub(crate) struct LastVotedForkSlotsEpochInfo {
+    pub epoch: Epoch,
+    pub total_stake: u64,
+    // Percentage (0 to 100) of total stake of active peers in this epoch,
+    // no matter they voted for a slot in this epoch or not.
+    pub voted_percent: f64,
+    // Percentage (0 to 100) of total stake of active peers which has voted for
+    // a slot in this epoch.
+    pub voted_for_this_epoch_percent: f64,
```

Comment:
I think it's better to track the actively voting stake, and convert to percent when we need to by dividing by the total stake

Reply:
Done.